### PR TITLE
[nrfconnect] Update nRF Connect SDK version to 2.4.0

### PR DESCRIPTION
**Pull request by kkasperczyk-no**
Regular update of nRF Connect SDK version to 2.4.0 including:

- recommended version bump
- docker image number update in github workflows
- nrfconnect platform bug fixes and improvements
- Zephyr platform alignment related to new Zephyr API

Depends on: https://github.com/project-chip/connectedhomeip/pull/27004

### CODE REVIEW

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 1]
```diff
@@ -3313,8 +3313,8 @@ bool DoorLockServer::HandleRemoteLockOperation(chip::app::CommandHandler * comma
 
     EndpointId endpoint       = commandPath.mEndpointId;
     OperationErrorEnum reason = OperationErrorEnum::kUnspecified;
-    Nullable<uint16_t> pinUserIdx; // Will get set to non-null if we find a user for the PIN.
-    Optional<uint16_t> pinCredIdx; // Will get set to a value if the PIN is one we know about.
+    Nullable<uint16_t> pinUserIdx;                                  // Will get set to non-null if we find a user for the PIN.
+    Optional<uint16_t> pinCredIdx{ Optional<uint16_t>::Missing() }; // Will get set to a value if the PIN is one we know about.
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:3317]: Why this change?  The default constructor initializes with Missing already.

File: src/inet/UDPEndPointImplSockets.cpp [hunk 2]
```diff
@@ -30,14 +30,20 @@
 #include <lib/support/SafeInt.h>
 #include <lib/support/logging/CHIPLogging.h>
 
+#if CHIP_SYSTEM_CONFIG_USE_POSIX_SOCKETS
 #if HAVE_SYS_SOCKET_H
 #include <sys/socket.h>
-#endif // HAVE_SYS_SOCKET_H
-
-#include <cerrno>
+#endif
```
Comment (bzbarsky-apple) [inline src/inet/UDPEndPointImplSockets.cpp:36]:
```suggestion
#endif // HAVE_SYS_SOCKET_H


```

File: .github/workflows/examples-nrfconnect.yaml [hunk 3]
```diff
@@ -40,7 +40,7 @@ jobs:
         if: github.actor != 'restyled-io[bot]'
 
         container:
-            image: connectedhomeip/chip-build-nrf-platform:0.7.3
+            image: connectedhomeip/chip-build-nrf-platform:0.7.16
```
Comment (andy31415) [inline .github/workflows/examples-nrfconnect.yaml:43]: you must change runners in integrations/cloudbuild as well, since those run nrf compilations as well.

In those files, please change all instances not just NRF, to avoid downloading 20+GB images several times.

File: config/nrfconnect/chip-module/CMakeLists.txt [hunk 4]
```diff
@@ -50,21 +50,12 @@ include(${COMMON_CMAKE_SOURCE_DIR}/chip_gn_args.cmake)
 include(${COMMON_CMAKE_SOURCE_DIR}/chip_gn.cmake)
 
 # Prepare compiler flags
-matter_add_flags(-D_DEFAULT_SOURCE)
-
-if (CHIP_CFLAGS)
```
Comment (Damian-Nordic) [inline config/nrfconnect/chip-module/CMakeLists.txt:55]: Is this removal intentional?

File: config/zephyr/Kconfig [hunk 5]
```diff
@@ -22,14 +22,14 @@ menuconfig CHIP
 	imply REQUIRES_FULL_LIBC
 	imply NEWLIB_LIBC_NANO
 	imply CBPRINTF_LIBC_SUBSTS
-	imply POSIX_API if !ARCH_POSIX
```
Comment (Damian-Nordic) [inline config/zephyr/Kconfig:25]: I'm not sure if Telink also wants to switch from `POSIX_API` to `NET_SOCKETS_POSIX_NAMES` so maybe it's better to make it conditional on our platform?
Unless @s07641069 confirms that Telink does want to follow this direction :)

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 1 above)
Comment (kkasperczyk-no) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:3317]: Right, however it seems some compilers are missing to see it is already initialized and generate the following warning:

> [709/772] Building CXX object CMakeFiles/app.dir/home/kamilkasperczyk/matter/connectedhomeip/src/app/clusters/door-lock-server/door-lock-server.cpp.obj
> /home/kamilkasperczyk/matter/connectedhomeip/src/app/clusters/door-lock-server/door-lock-server.cpp: In member function 'bool DoorLockServer::HandleRemoteLockOperation(chip::app::CommandHandler*, const chip::app::ConcreteCommandPath&, chip::app::Clusters::DoorLock::LockOperationTypeEnum, RemoteLockOpHandler, const chip::Optional<chip::Span<const unsigned char> >&)':
> /home/kamilkasperczyk/matter/connectedhomeip/src/app/clusters/door-lock-server/door-lock-server.cpp:3432:38: warning: 'pinCredIdx.chip::Optional<short unsigned int>::mValue.chip::Optional<short unsigned int>::Value::mData' may be used uninitialized [-Wmaybe-uninitialized]
>  3432 |         foundCred[0].credentialIndex = pinCredIdx.Value();
>       |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
> /home/kamilkasperczyk/matter/connectedhomeip/src/app/clusters/door-lock-server/door-lock-server.cpp:3317:24: note: 'pinCredIdx.chip::Optional<short unsigned int>::mValue.chip::Optional<short unsigned int>::Value::mData' was declared here
>  3317 |     Optional<uint16_t> pinCredIdx; //{ Optional<uint16_t>::Missing() }; // Will get set to a value if the PIN is one we know about.

File: config/nrfconnect/chip-module/CMakeLists.txt (hunk 4 above)
Comment (kkasperczyk-no) [reply config/nrfconnect/chip-module/CMakeLists.txt:55]: Discussed F2F and brought the change back.

File: config/zephyr/Kconfig (hunk 5 above)
Comment (kkasperczyk-no) [reply config/zephyr/Kconfig:25]: You are right, I added new kconfig to specify that nRF platform is built and implied it conditionally.

File: .github/workflows/examples-nrfconnect.yaml (hunk 3 above)
Comment (kkasperczyk-no) [reply .github/workflows/examples-nrfconnect.yaml:43]: Ok, applied, please re-review.

**Review by Damian-Nordic — APPROVED** [review]

**Comment by andy31415** [comment]
> chef

@erwinpan1 maybe. 
How does nrf in the regular builds decide what compiler to use? I am wondering if this is a case of using the nrf tools vs pigweed defaults.

**Comment by Damian-Nordic** [comment]
@kkasperczyk-no I have a patch to use the Zephyr toolchain for chef (someone asked to help with that in the past but I forgot to contribute it). Ping me on Monday about that :)

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 1 above)
Comment (andy31415) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:3317]: Compiler seems to not understand that `HasValue()` implies that `Value()` is valid. I wonder if there is a way to tell that, however I would rather have the fix be a `// NOLINT` or equivalent in the line using `Value` instead of this workaround.

Comment (bzbarsky-apple) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:3317]: Yeah, it does not make sense that the compiler "understands" that this copy constructor initializes things correctly but the default constructor does not.  In particular, the copy constructor "initializes" things to garbage bytes... ;)

We already have `#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"` in some places.  Perhaps we need it in a few more (like the copy constructor of Optional or something)?

**Review by LuDuda — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]
